### PR TITLE
New Turing-libtask integration 

### PR DESCRIPTION
**Pull request by yebai**
This PR introduces a simpler way of evaluating models. It is to avoid the current complicated calling stack encountered in #1744. The benefit is that we no longer need to trace many boiler-plate functions such as `evaluate!!`. For example,  `model(...) ==> evaluate!!  ==> _evaluate!!`  becomes 

```julia

julia> using Turing, DynamicPPL, AdvancedPS

julia> @model gdemo(x, y) = begin
         # Assumptions
         σ ~ InverseGamma(2,3)
         μ ~ Normal(0,sqrt(σ))
         # Observations
         x ~ Normal(μ, sqrt(σ))
         y ~ Normal(μ, sqrt(σ))
       end
gdemo (generic function with 2 methods)

# Case 1: Sample from the prior. 

julia> m = Turing.Core.TracedModel(gdemo(1.5, 2.), SampleFromPrior(), VarInfo()) 

julia> f = m.evaluator[1];

julia> args = m.evaluator[2:end];

julia> f(args...) 
(2.0, VarInfo (2 variables (μ, σ), dimension 2; logp: -6.162))


using Libtask

t = Libtask.CTask(f, args...) 

consume(t) # work fine!


# Case 2: SMC sampler 

julia> m = Turing.Core.TracedModel(gdemo(1.5, 2.), Sampler(SMC(50)), VarInfo());

julia> t = Libtask.CTask(m.evaluator[1], m.evaluator[2:end]...);

julia> consume(t)
ERROR: KeyError: key :__trace not found
Stacktrace:
 [1] consume(ttask::TapedTask)
   @ Libtask ~/.julia/dev/Libtask/src/tapedtask.jl:164
 [2] top-level scope
   @ REPL[58]:1


```

So, with this PR, the integration between Turing-Libtask should be exactly the same as that for AdvancedPS-Libtask.

<details>
  <summary>More tests (not needed until above example is fixed) </summary>

```julia
using Turing, Test, AbstractMCMC, DynamicPPL, Random

import AbstractMCMC.AbstractSampler

function check_numerical(chain,
    symbols::Vector,
    exact_vals::Vector;
    atol=0.2,
    rtol=0.0
)
    for (sym, val) in zip(symbols, exact_vals)
    E = val isa Real ?
    mean(chain[sym]) :
    vec(mean(chain[sym], dims=1))
    @info (symbol=sym, exact=val, evaluated=E)
    @test E ≈ val atol=atol rtol=rtol
    end
end

function check_MoGtest_default(chain; atol=0.2, rtol=0.0)
    check_numerical(chain,
        [:z1, :z2, :z3, :z4, :mu1, :mu2],
        [1.0, 1.0, 2.0, 2.0, 1.0, 4.0],
        atol=atol, rtol=rtol)
end

@model gdemo_d(x, y) = begin
    s ~ InverseGamma(2, 3)
    m ~ Normal(0, sqrt(s))
    x ~ Normal(m, sqrt(s))
    y ~ Normal(m, sqrt(s))
    return s, m
end

alg = CSMC(15)
chain = sample(gdemo_d(1.5, 2.0), alg, 5_00)
check_numerical(chain, [:s, :m], [49/24, 7/6], atol=0.1)
  
@model MoGtest(D) = begin
    mu1 ~ Normal(1, 1)
    mu2 ~ Normal(4, 1)
    z1 ~ Categorical(2)
    if z1 == 1
        D[1] ~ Normal(mu1, 1)
    else
        D[1] ~ Normal(mu2, 1)
    end
    z2 ~ Categorical(2)
    if z2 == 1
        D[2] ~ Normal(mu1, 1)
    else
        D[2] ~ Normal(mu2, 1)
    end
    z3 ~ Categorical(2)
    if z3 == 1
        D[3] ~ Normal(mu1, 1)
    else
        D[3] ~ Normal(mu2, 1)
    end
    z4 ~ Categorical(2)
    if z4 == 1
        D[4] ~ Normal(mu1, 1)
    else
        D[4] ~ Normal(mu2, 1)
    end
    z1, z2, z3, z4, mu1, mu2
end
  
MoGtest_default = MoGtest([1.0 1.0 4.0 4.0])

gibbs = Gibbs(PG(15, :z1, :z2, :z3, :z4), HMC(0.15, 3, :mu1, :mu2))
chain = sample(MoGtest_default, gibbs, 5_00);
check_MoGtest_default(chain, atol=0.15)
```

</details>

### CODE REVIEW

File: .github/workflows/TuringCI.yml [hunk 1]
```diff
@@ -52,6 +52,29 @@ jobs:
...
+            Pkg.update()
+            # Pkg.test()  # resolver may fail with test time deps
+          catch err
+            err isa Pkg.Resolve.ResolverError || rethrow()
+            # If we can't resolve that means this is incompatible by SemVer and this is fine
+            # It means we marked this as a breaking change, so we don't need to worry about
+            # Mistakenly introducing a breaking change, as we have intentionally made one
+            @info "Not compatible with this release. No problem." exception=err
+            exit(0)  # Exit immediately, as a success
+          end
```
Comment (yebai) [inline .github/workflows/TuringCI.yml:77]:
```suggestion
```

**Comment by yebai** [comment]
@KDr2  Could you take a look at why `case 2` is broken?

**Comment by KDr2** [comment]
> ```julia
> t = Libtask.CTask(m.evaluator[1], m.evaluator[2:end]...);
> ```

Below is the real bt of the error:

```
┌ Error: TapedTask Error:
│   exception =
│    KeyError: key :__trace not found
│    Stacktrace:
│      [1] getindex(d::IdDict{Any, Any}, key::Any)
│        @ Base ./iddict.jl:108
│      [2] current_trace()
│        @ AdvancedPS ~/.julia/packages/AdvancedPS/XPI6n/src/container.jl:64
│      [3] assume(rng::Random._GLOBAL_RNG, spl::Sampler{SMC{(), AdvancedPS.ResampleWithESSThreshold{typeof(AdvancedPS.resample_systematic), Int64}}}, dist::InverseGamma{Float64}, vn::VarName{:σ, Setfield.IdentityLens}, #unused#::UntypedVarInfo{DynamicPPL.Metadata{Dict{VarName, Int64}, Vector{Distribution}, Vector{VarName}, Vector{Real}, Vector{Set{DynamicPPL.Selector}}}, Float64})
│        @ Turing.Inference ~/Work/julia/Turing.jl/src/inference/AdvancedSMC.jl:327
│      [4] tilde_assume
│        @ ~/.julia/packages/DynamicPPL/c8MjC/src/context_implementations.jl:49 [inlined]
│      [5] tilde_assume
│        @ ~/.julia/packages/DynamicPPL/c8MjC/src/context_implementations.jl:46 [inlined]
│      [6] tilde_assume
│        @ ~/.julia/packages/DynamicPPL/c8MjC/src/context_implementations.jl:31 [inlined]
│      [7] tilde_assume!!(context::SamplingContext{Sampler{SMC{(), AdvancedPS.ResampleWithESSThreshold{typeof(AdvancedPS.resample_systematic), Int64}}}, DefaultContext, Random._GLOBAL_RNG}, right::InverseGamma{Float64}, vn::VarName{:σ, Setfield.IdentityLens}, vi::UntypedVarInfo{DynamicPPL.Metadata{Dict{VarName, Int64}, Vector{Distribution}, Vector{VarName}, Vector{Real}, Vector{Set{DynamicPPL.Selector}}}, Float64})
│        @ DynamicPPL ~/.julia/packages/DynamicPPL/c8MjC/src/context_implementations.jl:117
│      [8] (::Libtask.Instruction{typeof(Core._apply_iterate)})()
│        @ Libtask ~/Work/julia/Libtask.jl/src/tapedfunction.jl:84
│      [9] step_in(t::Libtask.Tape, args::Tuple{Model{typeof(gdemo), (:x, :y), (), (), Tuple{Float64, Float64}, Tuple{}, DefaultContext}, UntypedVarInfo{DynamicPPL.Metadata{Dict{VarName, Int64}, Vector{Distribution}, Vector{VarName}, Vector{Real}, Vector{Set{DynamicPPL.Selector}}}, Float64}, SamplingContext{Sampler{SMC{(), AdvancedPS.ResampleWithESSThreshold{typeof(AdvancedPS.resample_systematic), Int64}}}, DefaultContext, Random._GLOBAL_RNG}, Float64, Float64})
│        @ Libtask ~/Work/julia/Libtask.jl/src/tapedtask.jl:66
│     [10] (::Libtask.var"#11#12"{Libtask.TapedFunction, Tuple{Model{typeof(gdemo), (:x, :y), (), (), Tuple{Float64, Float64}, Tuple{}, DefaultContext}, UntypedVarInfo{DynamicPPL.Metadata{Dict{VarName, Int64}, Vector{Distribution}, Vector{VarName}, Vector{Real}, Vector{Set{DynamicPPL.Selector}}}, Float64}, SamplingContext{Sampler{SMC{(), AdvancedPS.ResampleWithESSThreshold{typeof(AdvancedPS.resample_systematic), Int64}}}, DefaultContext, Random._GLOBAL_RNG}, Float64, Float64}, Channel{Int64}, Channel{Any}})()
│        @ Libtask ./task.jl:123
└ @ Libtask ~/Work/julia/Libtask.jl/src/tapedtask.jl:31

```

**Comment by KDr2** [comment]
@yebai Do we need to copy the local storage of tasks when we copy CTasks?

**Comment by yebai** [comment]
I found the issue - we need to construct an `AdvancedPS.Trace` object so that task storage is correctly initialised with a backward trace reference. With a bit changes to AdvancedPS, I can now run `case 2` successfully. 

```julia

julia> using Turing, DynamicPPL, AdvancedPS, Libtask

julia> @model gdemo(x, y) = begin
         # Assumptions
         σ ~ InverseGamma(2,3)
         μ ~ Normal(0,sqrt(σ))
         # Observations
         x ~ Normal(μ, sqrt(σ))
         y ~ Normal(μ, sqrt(σ))
       end
gdemo (generic function with 2 methods)

julia> function AdvancedPS.Trace(f)
           if f isa Function
               ctask = Libtask.CTask(f)
           else
               # println(f.evaluator)
               # CTask(f, args) seems buggy, so we still use CTask(f::TracedModel) for now.
               # TODO: fix this after Libtask bug is resolved.
               ctask = Libtask.CTask(f.evaluator[1], f.evaluator[2:end]...)
               # ctask = Libtask.CTask(f)
           end

           # add backward reference
           newtrace = AdvancedPS.Trace(f, ctask)
           AdvancedPS.addreference!(ctask.task, newtrace)

           return newtrace
       end

julia> m = Turing.Core.TracedModel(gdemo(1.5, 2.), Sampler(SMC(50)), VarInfo());

julia> t = AdvancedPS.Trace(m);

julia> consume(t.ctask)
-4.143262128020936

```

**Comment by KDr2** [comment]
> Now the remaining pieces of complete Turing-Libtask integration are to make sure `t::Advanced.Trace` can be copied (i.e. `fork` and `forkr` work correctly), then the original and copied traces can be consumed, in a similar way to `CTask`.

I checked the old version of Libtask, it copies the local storage when a copy occurs. This commit added the same behavior: https://github.com/TuringLang/Libtask.jl/pull/100/commits/2fd124c62021632e8a7c3e203ceed8fd7d0aa32e

Can this ensure the `Trace` object to be correctly copied?

**Comment by yebai** [comment]
The code now runs correctly locally using https://github.com/TuringLang/AdvancedPS.jl/pull/40 and https://github.com/TuringLang/Libtask.jl/pull/102. More numerical fixes might be needed though.

**Comment by rikhuijzer** [comment]
@yebai In response to 915dfe0, you can temporarily add `Pkg.add("AdvancedPS.jl#hg/use-new-evaluation-api")` in the CI workflow to run a unregistered version in CI

**Comment by yebai** [comment]
@KDr2 There are only 4 errors left. After further debugging, I found that 3 errors are due to unsupported control flow, which can be ignored now. However, one error can not be attributed to known factors. It might be related to some unknown bug in the `intercept` function. Could you look into it, please? Here is an MWE. 

```julia

julia> using Turing, Test, AbstractMCMC, DynamicPPL, Random, Turing.RandomMeasures

julia> @model infiniteGMM(x) = begin
           # Hyper-parameters, i.e. concentration parameter and parameters of H.
           α = 1.0
           μ0 = 0.0
           σ0 = 1.0
           
           # Define random measure, e.g. Dirichlet process.
           rpm = DirichletProcess(α)
           
           # Define the base distribution, i.e. expected value of the Dirichlet process.
           H = Normal(μ0, σ0)
           
           # Latent assignment.
           z = tzeros(Int, length(x))
               
           # Locations of the infinitely many clusters.
           μ = tzeros(Float64, 0)
           
           for i in 1:length(x)
               
               # Number of clusters.
               K = maximum(z)
               nk = Vector{Int}(map(k -> sum(z .== k), 1:K))

               # Draw the latent assignment.
               z[i] ~ ChineseRestaurantProcess(rpm, nk)
               
               # Create a new cluster?
               if z[i] > K
                   push!(μ, 0.0)

                   # Draw location of new cluster.
                   μ[z[i]] ~ H
               end
                       
               # Draw observation.
               x[i] ~ Normal(μ[z[i]], 1.0)
           end
       end
infiniteGMM (generic function with 2 methods)

julia> # Generate some test data.
       Random.seed!(1);

julia> data = vcat(randn(10), randn(10) .- 5, randn(10) .+ 10);

julia> data .-= mean(data);

julia> data /= std(data);

julia> # MCMC sampling
       Random.seed!(2);

julia> iterations = 500;

julia> model_fun = infiniteGMM(data);

julia> chain = sample(model_fun, SMC(), iterations);
Sampling 100%
ERROR: TypeError: in new, expected Core.Box, got a value of type Libtask.Box{Core.Box}
Stacktrace:
  [1] macro expansion
    @ ./REPL[2]:23 [inlined]
  [2] macro expansion
    @ ~/.julia/packages/DynamicPPL/c8MjC/src/compiler.jl:529 [inlined]
  [3] ##300
    @ ./REPL[2]:19 [inlined]


```

**Comment by KDr2** [comment]
> ERROR: TypeError: in new, expected Core.Box, got a value of type Libtask.Box{Core.Box}

I am on it already, it is caused by the `new` operator, I haven't figured out a way to put the `new` operator in an instruction yet...

File: src/essential/container.jl [hunk 2]
```diff
@@ -1,19 +1,43 @@
...
+    # 1. Set the leaf context of `model.context` to `leafcontext(context)`.
+    # 2. Set leaf context of `context` to the context resulting from (1).
+    # The result is:
+    # `context` -> `childcontext(context)` -> ... -> `model.context`
+    #  -> `childcontext(model.context)` -> ... -> `leafcontext(context)`
+    return quote
+        context_new = DynamicPPL.setleafcontext(
+            context, DynamicPPL.setleafcontext(model.context, DynamicPPL.leafcontext(context))
+        )
+        (model.f, model, varinfo, context_new, $(unwrap_args...))
```
Comment (yebai) [inline src/essential/container.jl:38]:
```suggestion
        (model.f, model, DynamicPPL.resetlogp!!(varinfo), context_new, $(unwrap_args...))
```

**Comment by yebai** [comment]
I fixed the last CI error in 7c81350. Local tests are passing now - hopefully CI will also pass now!

**Comment by yebai** [comment]
All CI tests run without errors now - with https://github.com/TuringLang/Libtask.jl/pull/107, we should have full support for Julia 1.7 soon.

**Pull request merged**